src/java.base/macosx/classes/sun/nio/fs/MacOSXWatchService.java

```diff
@@ -0,0 +1,634 @@
+
+        // Native FSEventStreamRef as returned by FSEventStreamCreate().
+        private long         eventStreamRef;
+        private final Object eventStreamRefLock = new Object();
+
+        private DirectorySnapshot directorySnapshot;
+
+        MacOSXWatchKey(final MacOSXWatchService watchService,
+                       final UnixPath dir,
+                       final Object rootPathKey) throws IOException {
+            super(dir, watchService);
+            this.realRootPath       = dir.toRealPath().normalize();
+            this.realRootPathLength = realRootPath.toString().length() + 1;
+            this.rootPathKey        = rootPathKey;
+        }
+
+        synchronized void enable(final CFRunLoopThread runLoopThread,
+                                 final EnumSet<FSEventKind> eventsToWatch,
+                                 final EnumSet<WatchModifier> modifierSet) throws IOException {
+            assert(!isValid());
+
+            this.eventsToWatch = eventsToWatch;
+
+            directorySnapshot = DirectorySnapshot.create(getRealRootPath());
+
+            synchronized (eventStreamRefLock) {
+                final int kFSEventStreamCreateFlagWatchRoot  = 0x00000004;
+                eventStreamRef = MacOSXWatchService.eventStreamCreate(
+                        realRootPath.toString(),
+                        WatchModifier.sensitivityOf(modifierSet),
+                        kFSEventStreamCreateFlagWatchRoot);
+
+                if (eventStreamRef == 0)
+                    throw new IOException("Unable to create FSEventStream");
+
+                MacOSXWatchService.eventStreamSchedule(eventStreamRef,
+                        runLoopThread.getRunLoopRef());
+            }
+        }
+
+        synchronized void disable() {
+            invalidate();
+            directorySnapshot = null;
+        }
+
+        synchronized void handleEvents(final String[] paths, long eventFlagsPtr) {
+            if (paths == null) {
+                reportOverflow(null);
+                return;
+            }
+
+            if (updateNeeded(paths, eventFlagsPtr)) {
+                directorySnapshot.update(MacOSXWatchKey.this);
+            }
+        }
+
+        private Path toRelativePath(final String absPath) {
+            return   (absPath.length() > realRootPathLength)
+                    ? theFS.getPath(absPath.substring(realRootPathLength))
+                    : relativeRootPath;
+        }
+
+        private boolean updateNeeded(String[] paths, long eventFlagsPtr) {
+            boolean rootChanged = false;
+            for (final String absPath : paths) {
+                if (absPath == null) {
+                    reportOverflow(null);
+                    continue;
+                }
+
+                Path path = toRelativePath(absPath);
+
+                if (!relativeRootPath.equals(path)) {
+                    // Ignore events from subdirectories for now.
+                    continue;
```

Comment:
Should a line
```
eventFlagsPtr += SIZEOF_FS_EVENT_STREAM_EVENT_FLAGS;
```
be added here? (The definition of `SIZEOF_FS_EVENT_STREAM_EVENT_FLAGS` would have to be moved up.)

Reply:
Yes, most certainly. Thanks for catching this!